src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,213 @@
+			for (const language of this.cpdCatalog.rules[0].languages) {
+				const targetPaths: string[] = [];
+				for (const target of targets) {
+					for (const targetPath of target.paths) {
+						// when using --filelist, CPD will try to parse files not associated with language
+						// for apex, we only want to add .cls and .trigger files and for vf, we only want to look at .page and .component files
+						if (this.validateFileType(language, targetPath)) {
+							targetPaths.push(targetPath);
+						}
+					}
+				}
+
+				if (targetPaths.length === 0) {
+					this.logger.trace(`No matching cpd ${language} target files found. Nothing to execute.`);
+					return [];
+				}
+				this.logger.trace(`About to run CPD (${language}). Targets: ${targetPaths.length}`);
+
+				const selectedTargets = targetPaths.join(',');
+				const stdout = await CpdWrapper.execute(selectedTargets, language, await this.config.getMinimumTokens(ENGINE.CPD));
+				results.push(...this.processStdOut(stdout));
+				
+				this.logger.trace(`Found ${results.length} for CPD`);
+			}   
+			return results;
+		} catch (e) {
+			this.logger.trace('Cpd evaluation failed: ' + (e.message || e));
+			throw new SfdxError(e.message || e);
+		}
+	}
+
+	private validateFileType(language: string, fileName: string): boolean {
+		if (this.languageToPathMap.has(language)){
+			for (const fileType of this.languageToPathMap.get(language)) {
+				if (fileName.endsWith(fileType)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	protected processStdOut(stdout: string): RuleResult[] {
+		let violations: RuleResult[] = [];
+
+		this.logger.trace(`Output received from CPD: ${stdout}`);
+
+		const cpdEnd = '</pmd-cpd>';
+		const xmlStart = stdout.indexOf('<?xml');
+		const xmlEnd = stdout.lastIndexOf(cpdEnd);
+		if (xmlStart != -1 && xmlEnd != -1) {
+			const cpdXml = stdout.slice(xmlStart, xmlEnd + cpdEnd.length);
+			const cpdJson = xml2js(cpdXml, {compact: false, ignoreDeclaration: true}); 
+
+			const duplications =  cpdJson.elements[0].elements;
+			if (duplications) {
+				violations = this.jsonToRuleResults(duplications);
+			}
+		}
+
+		if (violations.length > 0) {
+			this.logger.trace('Found rule violations.');
+		}
+
+		return violations;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private jsonToRuleResults(duplications: any): RuleResult[] {    
+		const ruleResults: RuleResult[] = [];
+
+		for (const duplication of duplications) {
+		
+			const parts = duplication.elements;
+			const codeFragment = parts.find(part => {return part.name === 'codefragment'}).elements[0].cdata;
+			const occurences = parts.filter(part => {return part.name === 'file'});
+			const codeFragmentID = checksum(codeFragment).slice(0,7);
+			// TODO: check for collisions (cpd can return multiple duplications for the same code fragment, so these have the same checksum currently)
```

Comment:
Not sure why, however for the same code fragment, cpd can create multiple duplications (don't know why they aren't grouped under one when it's the exact same code). This means we can have "a62b678: 1 of 2 duplications..." as well as "a62b678: 1 of 38 duplications...". They have the same checksum because all 38+2 instances are the same code fragment. Should we leave it this way because they are really the same duplication? If so, what about collisions, how do we tell if it a collision is due to the same code or different code fragments that have the same checksum?

Reply:
Would it help to prefix/suffix the length of characters in the code segment to get different IDs? If the difference between the two segments is whitespace and the length should help us distinguish them.

Reply:
The strings seem to be exactly the same.